src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -247,7 +248,7 @@ public SessionFactoryImpl(Configuration cfg, IMapping mapping, Settings settings
 					if (cache != null)
 					{
 						caches.Add(cacheRegion, cache);
-						allCacheRegions.Add(cache.RegionName, cache.Cache);
+						((IDictionary<string, ICache>)allCacheRegions).Add(cache.RegionName, cache.Cache);
```

Comment:
I think the explicit interface implementations of the `ConcurrentDictionary` are not thread safe (that's why they are explicitly implemented). I had a similar solution on my machine. Currently I'm in holidays. I'll compare yours to mine when I'm back.